experiment/coverage_utils.py

```diff
@@ -0,0 +1,206 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Utility functions for coverage report generation."""
+
+import os
+import multiprocessing
+import tarfile
+import posixpath
+
+from common import filesystem
+from common import experiment_utils
+from common import new_process
+from common import benchmark_utils
+from common import fuzzer_utils
+from common import logs
+from common import filestore_utils
+from experiment import measurer
+
+logger = logs.Logger('coverage_utils')  # pylint: disable=invalid-name
+
+
+def generate_cov_reports(experiments, benchmarks, fuzzers, report_dir):
+    """Generate coverage reports for each benchmark and fuzzer."""
+    logger.info('Start generating coverage report for benchmarks.')
+    set_up_coverage_files(experiments, report_dir, benchmarks)
+    with multiprocessing.Pool() as pool:
+        generate_cov_report_args = [(experiments, benchmark, fuzzer, report_dir)
+                                    for benchmark in benchmarks
+                                    for fuzzer in fuzzers]
+        pool.starmap(generate_cov_report, generate_cov_report_args)
+        pool.close()
+        pool.join()
+    logger.info('Finished generating coverage report.')
+
+
+def generate_cov_report(experiments, benchmark, fuzzer, report_dir):
+    """Generate the coverage report for one pair of benchmark and fuzzer."""
+    logs.initialize()
+    logger.info('Generating coverage report for benchmark: {benchmark} \
+                fuzzer: {fuzzer}.'.format(benchmark=benchmark, fuzzer=fuzzer))
+    generator = CoverageReporter(fuzzer, benchmark, experiments, report_dir,
+                                 logger)
+    # Gets and merges all the profdata files.
+    generator.fetch_profdata_file()
+    generator.merge_profdata_files()
+    # Generates the reports using llvm-cov.
+    generator.generate_cov_report()
+
+    logger.info('Finished generating coverage report for '
+                'benchmark:{benchmark} fuzzer:{fuzzer}.'.format(
+                    benchmark=benchmark, fuzzer=fuzzer))
+
+
+def set_up_coverage_files(experiment_names, report_dir, benchmarks):
+    """Sets up coverage files for all benchmarks."""
+    with multiprocessing.Pool() as pool:
```

Comment:
No need of multiprocessing here at all, these gcs file cp is fast, remove complexity.